core/src/main/scala/org/bykn/bosatsu/codegen/clang/ClangGen.scala

```diff
@@ -1022,10 +1027,20 @@ object ClangGen {
           case Local(arg) =>
             directFn(arg)
               .flatMap {
-                case Some((nm, false, arity)) =>
-                  // a closure can't be a static name
-                  pv(boxFn(nm, arity))
-                case _ =>
+                case Some((nm, isClosure, arity)) =>
+                  if (!isClosure) {
+                    // a closure can't be a static name
+                    pv(boxFn(nm, arity))
+                  }
+                  else {
+                    // we need to allocate another wrapper
```

Comment:
this was a bug... if we have a direct function that is a closure, we can't just pass it by value, we have to box it in a closure wrapper to carry the environment.